v2/cli/init_command.go

```diff
@@ -0,0 +1,243 @@
+	if err = checkFileOverwrite(
+		path.Join(".brigade", "project.yaml"),
+		buf,
+	); err != nil {
+		return err
+	}
+
+	// TODO: Right now the notes doesn't have template fields, add later!
+	notesBuf, err :=
+		execTemplate(notesTemplate, fields)
+	if err != nil {
+		return err
+	}
+
+	// Create a NOTES.txt file
+	if err = checkFileOverwrite(
+		path.Join(".brigade", "NOTES.txt"),
+		notesBuf,
+	); err != nil {
+		return err
+	}
+
+	secretsBuf, err :=
+		execTemplate(secretsTemplate, fields)
+	if err != nil {
+		return err
+	}
+
+	// Create a secrets.yaml file
+	if err = checkFileOverwrite(
+		path.Join(".brigade", "secrets.yaml"),
+		secretsBuf,
+	); err != nil {
+		return err
+	}
+
+	err = addToGitignore(".brigade/secrets.yaml", "./.gitignore")
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Detect if .gitignore exists. If it does, add .brigade/secrets.yaml to it
+// If not, create it and add .brigade/secrets.yaml to it unconditionally
+func addToGitignore(pathToIgnore string, gitIgnorePath string) error {
+	if fileExists, err := file.Exists(gitIgnorePath); err != nil {
+		return err
+	} else if fileExists {
+		f, err := os.OpenFile(
+			gitIgnorePath,
+			os.O_APPEND|os.O_WRONLY|os.O_CREATE,
+			0600,
+		)
+		if err != nil {
+			return err
+		}
+
+		defer f.Close()
+
+		// Add path to .gitignore unconditionally
+		if _, err = f.WriteString("\n" + pathToIgnore); err != nil {
+			return err
+		}
+	} else {
+		fmt.Println("Creating .gitignore...")
+		if err := ioutil.WriteFile(
+			gitIgnorePath,
+			[]byte(pathToIgnore),
+			0644,
+		); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func idMatch(id string) error {
+	regex := regexp.MustCompile(`^[a-z][a-z\d-]*[a-z\d]$`)
+	idMatch := regex.MatchString(id)
+	if !idMatch || len(id) < 3 || len(id) > 63 {
+		return fmt.Errorf("invalid value %q for --id flag"+
+			" (3-63 characters, first char must be"+
+			" a letter, lowercase letters only)", id)
+	}
+	return nil
+}
+
+func normalizeLanguage(language string) (string, error) {
+	switch language {
+	case "ts", "typescript":
+		return "ts", nil
+	case "js", "javascript":
+		return "js", nil
+	default:
+		return "", fmt.Errorf("unrecognized value %q for --language flag"+
+			"(ts or js expected)", language)
+	}
+}
+
+// Check if the given file exists, and provide
+// confirmation to overwrite that file
+func checkFileOverwrite(
+	filepath string,
+	buf bytes.Buffer,
+) error {
+	if fileExists, err := file.Exists(filepath); err != nil {
+		return err
+	} else if fileExists {
+		response := false
+		prompt := &survey.Confirm{
+			Message: "A " + path.Base(filepath) + " file already exists in your" +
+				" .brigade directory, would you like to overwrite it?",
+			Default: false,
+		}
+
+		if err = survey.AskOne(prompt, &response); err != nil {
+			return err
+		}
+
+		if !response {
+			return nil
+		}
+	}
+
+	if err := ioutil.WriteFile(filepath, buf.Bytes(), 0644); err != nil {
+		return err
+	}
+	return nil
+}
```

Comment:
I think this function does too much stuff, which makes it difficult to test and even more difficult to reuse bits of it in the future.

Consider having this function only be responsible for determining if the file already exists, prompting for overwrite confirmation if necessary, and either way, returning a boolean that simply tells the caller whether they can proceed with writing (or overwriting, as the case may be) the file.

This way the concerns of validating user intent and actually writing the file are cleanly separated from one another and can be tested independently of one another.

Reply:
Pushed a change that makes this function return a boolean for recording the user response, and moved the logic for writing the file out of the function. Let me know if that's the right idea!


	if overwrite, err := checkFileOverwrite(
		path.Join(".brigade", "NOTES.txt"),
		notesBuf,
	); err != nil {
		return err
	} else if overwrite {
		if err := ioutil.WriteFile(
			path.Join(
				".brigade", "NOTES.txt"),
			notesBuf,
			0644,
		); err != nil {
			return err
		}
	}

Reply:
This is, indeed the right idea. Try to do the path. Join stuff just once tho... save it in a variable. It's always good to write things in a way that if you have to change something someday, it's done in just one place.